File: orchestrator/src/trace.ts
import { convex, api } from "./convexClient.js";
import type { Id } from "../../convex-backend/convex/_generated/dataModel.js";
import { costUsd } from "./config.js";

type TraceType =
  | "manager_plan" | "delegate" | "specialist" | "subspecialist"
  | "llm_call" | "linkup_query" | "dedup" | "score" | "synthesis"
  | "escalation" | "error";

function truncate(s: unknown, n = 4000): string | undefined {
  if (s == null) return undefined;
  const str = typeof s === "string" ? s : JSON.stringify(s);
  return str.length > n ? str.slice(0, n) + "…[truncated]" : str;
}

// Tracer owns the sequence counter and running cost/token totals for a run.
// Each span is a node in the trace tree (parentId links to its caller).
export class Tracer {
  private seq = 0;
  totalTokensIn = 0;
  totalTokensOut = 0;
  totalCostUsd = 0;

  constructor(public runId: Id<"runs">) {}

  async span(opts: {
    parentId?: Id<"traces">;
    agent: string;
    type: TraceType;
    label: string;
    status?: "ok" | "error" | "running";
    model?: string;
    tokensIn?: number;
    tokensOut?: number;
    latencyMs?: number;
    input?: unknown;
    output?: unknown;
    error?: string;
  }): Promise<Id<"traces">> {
    const tokensIn = opts.tokensIn ?? 0;
    const tokensOut = opts.tokensOut ?? 0;
    const cost = opts.model ? costUsd(opts.model, tokensIn, tokensOut) : 0;
    this.totalTokensIn += tokensIn;
    this.totalTokensOut += tokensOut;
    this.totalCostUsd += cost;

    return await convex.mutation(api.traces.emit, {
      runId: this.runId,
      parentId: opts.parentId,
      seq: this.seq++,
      agent: opts.agent,
      type: opts.type,
      label: opts.label,
      status: opts.status ?? "ok",
      model: opts.model,
      tokensIn,
      tokensOut,
      costUsd: cost,
      latencyMs: opts.latencyMs ?? 0,
      input: truncate(opts.input),
      output: truncate(opts.output),
      error: opts.error,
    });
  }

  // Wrap an async unit of work as a timed span. Records error spans on throw.
  async track<T>(
    meta: { parentId?: Id<"traces">; agent: string; type: TraceType; label: string; input?: unknown },
    fn: () => Promise<{ result: T; tokensIn?: number; tokensOut?: number; model?: string; output?: unknown }>,
  ): Promise<T> {
    const started = Date.now();
    try {
      const r = await fn();
      await this.span({
        ...meta,
        status: "ok",
        model: r.model,
        tokensIn: r.tokensIn,
        tokensOut: r.tokensOut,
        latencyMs: Date.now() - started,
        output: r.output,
      });
      return r.result;
    } catch (e) {
      await this.span({
        ...meta,
        status: "error",
        latencyMs: Date.now() - started,
        error: e instanceof Error ? e.message : String(e),
      });
      throw e;
    }
  }
}

File: agent/stalker/trace.py
"""Tracer — emits one trace event per agent/tool/LLM step to Postgres + Convex.
Owns the run's sequence counter and running token/cost totals. This is the
observability spine: the trace tree is built from (seq, parent_seq), which is
identical in both stores, so no cross-store id mapping is needed."""
from __future__ import annotations

import time
from typing import Any
from . import convex_client as cvx
from .config import cost_usd
from .db import session_scope
from .models import TraceEvent


def _truncate(v: Any, n: int = 4000) -> str | None:
    if v is None:
        return None
    s = v if isinstance(v, str) else str(v)
    return s[:n] + "…[truncated]" if len(s) > n else s


class Tracer:
    def __init__(self, run_pg_id: int | None, run_convex_id: str | None):
        self.run_pg_id = run_pg_id
        self.run_convex_id = run_convex_id
        self._seq = 0
        self.tokens_in = 0
        self.tokens_out = 0
        self.cost_usd = 0.0

    def span(
        self,
        agent: str,
        type: str,
        label: str,
        parent_seq: int | None = None,
        status: str = "ok",
        model: str | None = None,
        tokens_in: int = 0,
        tokens_out: int = 0,
        latency_ms: int = 0,
        input: Any = None,
        output: Any = None,
        error: str | None = None,
    ) -> int:
        """Record a span. Returns its seq (pass as a child's parent_seq)."""
        cost = cost_usd(model, tokens_in, tokens_out) if model else 0.0
        self.tokens_in += tokens_in
        self.tokens_out += tokens_out
        self.cost_usd += cost
        seq = self._seq
        self._seq += 1

        with session_scope() as s:
            if s is not None and self.run_pg_id is not None:
                s.add(TraceEvent(
                    run_id=self.run_pg_id, parent_seq=parent_seq, seq=seq, agent=agent,
                    type=type, label=label, status=status, model=model,
                    tokens_in=tokens_in, tokens_out=tokens_out, cost_usd=cost,
                    latency_ms=latency_ms, input=_truncate(input), output=_truncate(output),
                    error=error,
                ))

        if self.run_convex_id:
            cvx.mutation("traces:emit", {
                "runId": self.run_convex_id, "seq": seq, "parentSeq": parent_seq,
                "agent": agent, "type": type, "label": label, "status": status,
                "model": model, "tokensIn": tokens_in, "tokensOut": tokens_out,
                "costUsd": cost, "latencyMs": latency_ms,
                "input": _truncate(input), "output": _truncate(output), "error": error,
            })
        return seq


class Timer:
    """with Timer() as t: ...; t.ms → elapsed milliseconds."""
    def __enter__(self):
        self._start = time.time()
        return self

    def __exit__(self, *a):
        self.ms = int((time.time() - self._start) * 1000)
